Source/WebKit/Platform/CoroutineUtilities.h

```diff
@@ -131,55 +120,48 @@ class Lazy {
 
 struct Task {
     struct promise_type {
-        Task get_return_object() { return { std::coroutine_handle<promise_type>::from_promise(*this) }; }
+        Task get_return_object() { return { }; }
         std::suspend_never initial_suspend() { return { }; }
-        std::suspend_always final_suspend() noexcept { return { }; }
+        std::suspend_never final_suspend() noexcept { return { }; }
         void unhandled_exception() { }
         void return_void() { }
     };
-    std::coroutine_handle<promise_type> handle;
 };
 
-class CoroutineCaller {
-    WTF_MAKE_TZONE_ALLOCATED_INLINE(CoroutineCaller);
+template<typename> class Awaitable;
+template<typename T> class [[nodiscard]] Awaitable {
 public:
-    CoroutineCaller() = default;
-    void setCoroutine(Function<Task()>&& coroutine)
+    using Callback = CompletionHandler<void(CompletionHandler<void(T&&)>)>;
+    Awaitable(Callback&& callback)
+        : m_callback(WTFMove(callback)) { }
+    bool await_ready() { return !!m_result; }
```

Comment:
`std::optional<T>` has an `operator bool()` so I think it's safe to:

```
bool await_ready() const { return m_result; }
```

Reply:
also, make it `const`?

Reply:
const yes, !! needs to stay.  std::optional::operator bool is explicit.